File: tulip-standalone/mbench-jbang-java/HttpUser.java
/*-------------------------------------------------------------------------*/

import io.github.wfouche.tulip.api.TulipUser;
import io.github.wfouche.tulip.api.TulipUtils;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

/*-------------------------------------------------------------------------*/

public class HttpUser extends TulipUser {

    // ----------------------------------------------------------------- //

    private HttpRequest requestPosts = createRequest("posts");
    private HttpRequest requestComments = createRequest("comments");
    private HttpRequest requestAlbums = createRequest("albums");
    private HttpRequest requestPhotos = createRequest("photos");
    private HttpRequest requestTodos = createRequest("todos");

    private int userId;

    public HttpUser(int userId, int threadId) {
        super(userId, threadId);
        this.userId = userId;
    }

    public boolean onStart() {
        var actionId = 0;
        return true;
    }

    // ----------------------------------------------------------------- //

    public boolean action1() {
        // 6 ms delay (average)
        TulipUtils.delayMillisRandom(0, 12);
        return true;
    }

    public boolean action2() {
        // 14 ms delay (average)
        TulipUtils.delayMillisRandom(0, 28);
        return true;
    }

    // ----------------------------------------------------------------- //

    public boolean action3() {
        return serviceCall(requestPosts);
    }

    public boolean action4() {
        return serviceCall(requestComments);
    }

    public boolean action5() {
        return serviceCall(requestAlbums);
    }

    public boolean action6() {
        return serviceCall(requestPhotos);
    }

    public boolean action7() {
        return serviceCall(requestTodos);
    }

    // ----------------------------------------------------------------- //

    public boolean action8() {
        var actionId = 8;
        return true;
    }

    // ----------------------------------------------------------------- //

    public boolean action9()  {
        return true;
    }

    // ----------------------------------------------------------------- //

    public boolean action10() {
        TulipUtils.delayMillisFixed(10);
        return true;
    }

    // ----------------------------------------------------------------- //

    public boolean onStop() {
        //Thread.sleep(100)
        return true;
    }

    // ----------------------------------------------------------------- //

    private HttpRequest createRequest(String name) {
        try {
            var id = userId + 1;
            var url = this.getUserParamValue("url");
            var request = HttpRequest.newBuilder()
                    .uri(new URI(url + "/" + name + "/" + String.valueOf(id)))
                    .GET()
                    .build();
            return request;
        } catch (java.lang.Exception e) {
            throw new RuntimeException(e);
        }

    }

    private static HttpClient client = HttpClient.newHttpClient();

    private static boolean serviceCall(HttpRequest request) {
        try {
            var response = client.send(request, HttpResponse.BodyHandlers.ofString());

            //println(id)
            //println(name)
            //println(response.statusCode())
            //println(response.body())

            return (response.statusCode() == 200);
        } catch (java.lang.Exception e) {
            throw new RuntimeException(e);
        }
    }

}

/*-------------------------------------------------------------------------*/

File: tulip-main/src/main/kotlin/org/example/user/HttpUser.kt
package org.example.user

/*-------------------------------------------------------------------------*/

import io.github.wfouche.tulip.api.TulipUtils
import io.github.wfouche.tulip.api.TulipUser
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

import org.slf4j.Logger
import org.slf4j.LoggerFactory

/*-------------------------------------------------------------------------*/

private val client = HttpClient.newHttpClient()

private fun serviceCall(request:HttpRequest): Boolean {
    // https://www.baeldung.com/java-httpclient-connection-management

    val response = client.send(request, HttpResponse.BodyHandlers.ofString())

    //println(id)
    //println(name)
    //println(response.statusCode())
    //println(response.body())

    return (response.statusCode() == 200)
}

/*-------------------------------------------------------------------------*/

class HttpUser(userId: Int, threadId: Int) : TulipUser(userId, threadId) {

    // ----------------------------------------------------------------- //

    private val requestPosts = createRequest("posts")
    private val requestComments = createRequest("comments")
    private val requestAlbums = createRequest("albums")
    private val requestPhotos = createRequest("photos")
    private val requestTodos = createRequest("todos")

    override fun onStart(): Boolean {
        if (userId == 0) {
            var s = ""
            s = "debug: " + getUserParamValue("debug").toBoolean()
            logger.info(s)
            s = "http_port: " + getUserParamValue("http_port").toInt()
            logger.info(s)
        }
        return true
    }

    // ----------------------------------------------------------------- //

    // 0.25*6 + 0.75*14 = 12.0 ms

    override fun action1(): Boolean {
        // 6 ms delay (average)
        TulipUtils.delayMillisRandom(0, 12)
        return true
    }

    override fun action2(): Boolean {
        // 14 ms delay (average)
        TulipUtils.delayMillisRandom(0, 28)
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action3(): Boolean {
        return serviceCall(requestPosts)
    }

    override fun action4(): Boolean {
        return serviceCall(requestComments)
    }

    override fun action5(): Boolean {
        return serviceCall(requestAlbums)
    }

    override fun action6(): Boolean {
        return serviceCall(requestPhotos)
    }

    override fun action7(): Boolean {
        return serviceCall(requestTodos)
    }

    // ----------------------------------------------------------------- //

    override fun action8(): Boolean {
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action9(): Boolean {
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action10(): Boolean {
        TulipUtils.delayMillisFixed(10)
        return true
    }

    // ----------------------------------------------------------------- //

    override fun onStop(): Boolean {
        logger.info("  Terminate: UserId = $userId")
        Thread.sleep(100)
        return true
    }

    // ----------------------------------------------------------------- //

    private fun createRequest(name: String): HttpRequest {
        val id = userId + 1
        val url: String = this.getUserParamValue("url")
        val request:HttpRequest = HttpRequest.newBuilder()
            .uri(URI("${url}/${name}/${id}"))
            .GET()
            .build()
        return request
    }

    // ----------------------------------------------------------------- //

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(HttpUser::class.java)
    }

}

/*-------------------------------------------------------------------------*/
